### ISSUE TITLE
Pix2seq: A Language Modeling Framework for Object Detection

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#129]
## TL; DR

- Recast object detection as **image captioning** task, as the title *pix2seq* suggests.
- Given an image with diverse objects, prompt the model to *speak* the locations and object classes using Transformer decoder.
- Incorporate pixel-level discretization for bbox and sequence augmentation for better precision/recall trade-off.

## Motivation

- If a model has learned where and what about the objects in an image, no special architecture is needed; one just need to make it *speak*.

## Previous Work

- Faster-RCNN
- DETR

## Method

![image](https://user-images.githubusercontent.com/46308407/145022259-2314a059-1b93-496d-a6af-4fdbc5fbe9bf.png)

- Use a typical image encoder (e.g., ResNet-101) to encode image features.
- Use a transformer encoder (6-layer) + transformer decoder (6-layer) to output annotation sequence.
- Bbox coordinates and class labels share the same vocabulary (total #vocabs = n_bins + n_classes)
- Discretize the image pixel space using binning. For 600x600 image, 600 bins suffice, which is far fewer than typical language task.

![image](https://user-images.githubusercontent.com/46308407/145022904-2a9d2a9f-de5f-43cc-8dc1-ee2fa189e377.png)

### Objective

![image](https://user-images.githubusercontent.com/46308407/145022872-ce9a9cff-a54d-43bc-9b6e-9cb4d59c440b.png)

- Typical maximum likelihood objective

### Sequence Augmentation for Better Recall

- Pix2seq model without sequence augmentation typically show **low recall**, outputting <EOS> token when not all objects are detected.
- This can be because of annotation errors or uncertainty in recognizing/localizing certain objects.
- To achieve better trade off in precision/recall, generate noisy objects with "noise" labels, and encourage the model to predict "noise" classes for corresponding objects.
- This way, we can prompt the model to output a fixed-length sequence that potentially covers all objects, while minimizing the shortcomings of repeated or wrong predictions.
- This is a smart way of obtaining better recall.

![image](https://user-images.githubusercontent.com/46308407/145023271-23c1410c-b4a5-42a5-9136-2c59830622c1.png)

## Experiments

![image](https://user-images.githubusercontent.com/46308407/145023391-7558a88a-85d9-4dbe-9c15-4dec68e45427.png)
![image](https://user-images.githubusercontent.com/46308407/145023410-253bb5f2-d4f9-4a73-8828-7b304fa0275f.png)

- About 300 bins suffice in terms of AP.

![image](https://user-images.githubusercontent.com/46308407/145023454-e89c0ff5-0b9d-4537-82a9-f037c0a4b015.png)
![image](https://user-images.githubusercontent.com/46308407/145023511-a1ae867d-599c-4149-9b14-9a4f6d3b80d3.png)

## Comments

- Very interesting step towards a general AI. No task-specific architectural / objective function design is needed. *General understanding is all you need.*
- Once again, transformer has done its share of surprise.
- Object detection has traditionally been one of the research areas with the strongest pressure for engineering and tuning. How they simply reached near-SOTA performance with such a simple method is simply impressive.
- Clever get-away for low recall problem. Enforcing fixed length list output while teaching the model to distinguish noise classes was really smart.
